adbc_driver_manager.OperationalError: UNKNOWN: [Snowflake] arrow/ipc: unknown error while reading: cannot allocate memory

Hello,

Was trying the adbc python snow driver and I see something odd . Simple function 

```
# Query via adbc
def query_via_adbc_arrow(conn_adbc, query, return_type='arrow'):
    print(f"Enters query_via_adbc_arrow  ")
    start_time = datetime.datetime.now()
    cursor = conn_adbc.cursor()
    cursor.execute(query)
    # Return pandas or arrow
    if return_type == 'arrow':
        df = cursor.fetch_arrow_table()
        num_rows = df.num_rows
    else:
        df = cursor.fetch_df() 

        #df = cursor.fetch_arrow_table().to_pandas()
        num_rows = df.shape[0]
    cursor.close()
    end_time = datetime.datetime.now()
    execution_time = (end_time - start_time).total_seconds()
    print(f" Got {num_rows} rows  in {return_type} . The execution time is {execution_time} secs ")
    print(f"Exits query_via_adbc_arrow ")
```

Sometimes it works and some times it does not ( the regular snowflake python connector works every time ) 
When we run adbc part there are weird errors 
(310) [aborissov@bhsys-data-dev-euw1c-lnx10 python_project2]$ python3 snow_python_adbc.py
Enters query_via_adbc_arrow
 Got 10003143 rows  in arrow . The execution time is 1.62057 secs
Exits query_via_adbc_arrow
Enters query_via_adbc_arrow
Traceback (most recent call last):
  File "/<>/python_project2/snow_python_adbc.py", line 133, in <module>
    df = query_via_adbc_arrow(conn_adbc, "select * from private.main_td_limit_orders", return_type='pandas')
  File "/bh<>/snow_python_adbc.py", line 91, in query_via_adbc_arrow
    df = cursor.fetch_df() # Does not work well with large data running out of memory
  File "/<>/venvs/310/lib/python3.10/site-packages/adbc_driver_manager/dbapi.py", line 1050, in fetch_df
    return self._results.fetch_df()
  File "<>/venvs/310/lib/python3.10/site-packages/adbc_driver_manager/dbapi.py", line 1139, in fetch_df
    return self._reader.read_pandas()
  File "adbc_driver_manager/_reader.pyx", line 108, in adbc_driver_manager._reader.AdbcRecordBatchReader.read_pandas
  File "adbc_driver_manager/_reader.pyx", line 40, in adbc_driver_manager._reader._AdbcErrorHelper.check_error
adbc_driver_manager.OperationalError: UNKNOWN: [Snowflake] arrow/ipc: unknown error while reading: cannot allocate memory
(310) [<>python_project2]$
(310) [<> python_project2]$
(310) [<> python_project2]$
(310) [<> python_project2]$
(310) [<> python_project2]$
(310) [<>python_project2]$ python3 snow_python_adbc.py
Enters query_via_adbc_arrow
 Got 10003143 rows  in arrow . The execution time is 1.535479 secs
Exits query_via_adbc_arrow
Enters query_via_adbc_arrow
Traceback (most recent call last):
  File "/<>/python_project2/python_project2/snow_python_adbc.py", line 133, in <module>
    df = query_via_adbc_arrow(conn_adbc, "select * from private.main_td_limit_orders", return_type='pandas')
  File "/<>/python_project2/python_project2/snow_python_adbc.py", line 91, in query_via_adbc_arrow
    df = cursor.fetch_df() # Does not work well with large data running out of memory
  File "/<>/venvs/310/lib/python3.10/site-packages/adbc_driver_manager/dbapi.py", line 1050, in fetch_df
    return self._results.fetch_df()
  File "<>/venvs/310/lib/python3.10/site-packages/adbc_driver_manager/dbapi.py", line 1139, in fetch_df
    return self._reader.read_pandas()
  File "adbc_driver_manager/_reader.pyx", line 108, in adbc_driver_manager._reader.AdbcRecordBatchReader.read_pandas
  File "adbc_driver_manager/_reader.pyx", line 40, in adbc_driver_manager._reader._AdbcErrorHelper.check_error
adbc_driver_manager.OperationalError: UNKNOWN: [Snowflake] arrow/ipc: unknown error while reading: cannot allocate memory
(310) [<> python_project2]$
(310) [<> python_project2]$
(310) [<> python_project2]$
(310) [<> python_project2]$
(310) [<> python_project2]$ python3 snow_python_adbc.py
Enters query_via_adbc_arrow
 Got 10003143 rows  in arrow . The execution time is 1.560988 secs
Exits query_via_adbc_arrow
Enters query_via_adbc_arrow
 Got 10003143 rows  in pandas . The execution time is 2.706668 secs
Exits query_via_adbc_arrow
(310) [<> python_project2]$

Has anyone experienced that ?  Thanks

I've built from master (it installed `adbc_driver_snowflake` version '0.0.0+gff222e20' following the instructions [here](https://github.com/apache/arrow-adbc/tree/main/python/adbc_driver_snowflake#building)) and I'm still getting some errors, as the examples listed below:

```
panic: close of nil channel

goroutine 232 [running]:
github.com/apache/arrow-adbc/go/adbc/driver/snowflake.newRecordReader.func4()
        /adbc/go/adbc/driver/snowflake/record_reader.go:572 +0x85
created by github.com/apache/arrow-adbc/go/adbc/driver/snowflake.newRecordReader
        /adbc/go/adbc/driver/snowflake/record_reader.go:568 +0xbc8
```

or (with more info printed from `traceback.format_exc()`)
```
Traceback (most recent call last):
  File "/app/snowflake.py", line 41, in export_table
    Table = cursor.fetch_arrow_table()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/arrow-adbc/python/adbc_driver_manager/adbc_driver_manager/dbapi.py", line 1032, in fetch_arrow_table
    return self._results.fetch_arrow_table()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/arrow-adbc/python/adbc_driver_manager/adbc_driver_manager/dbapi.py", line 1135, in fetch_arrow_table
    return self._reader.read_all()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "adbc_driver_manager/_reader.pyx", line 90, in adbc_driver_manager._reader.AdbcRecordBatchReader.read_all
  File "adbc_driver_manager/_reader.pyx", line 40, in adbc_driver_manager._reader._AdbcErrorHelper.check_error
adbc_driver_manager.OperationalError: UNKNOWN: [Snowflake] arrow/ipc: unknown error while reading: cannot allocate memory
```


The tables that failed have 2.2M, 98K and 13.3K rows with 48.0MB, 41.9MB and 229.5KB, respectively.

Python packages:

- -e git+https://github.com/apache/arrow-adbc.git@ff222e2004e281d26395964d982320bf5c506160#egg=adbc_driver_manager&subdirectory=python/adbc_driver_manager
- -e git+https://github.com/apache/arrow-adbc.git@ff222e2004e281d26395964d982320bf5c506160#egg=adbc_driver_snowflake&subdirectory=python/adbc_driver_snowflake
- numpy==1.26.1
- pyarrow==14.0.1
- python-dotenv==1.0.0
- setuptools==68.2.2
- wheel==0.41.3

Please let me know if there's anything else I could do to help debugging this. I'd be happy to contribute.

--- a/go/adbc/driver/snowflake/record_reader.go
+++ b/go/adbc/driver/snowflake/record_reader.go
@@ -567,17 +567,6 @@ func newRecordReader(ctx context.Context, alloc memory.Allocator, ld gosnowflake
 
 	lastChannelIndex := len(chs) - 1
 	go func() {
-		// place this here so that we always clean up, but they can't be in a
-		// separate goroutine. Otherwise we'll have a race condition between
-		// the call to wait and the calls to group.Go to kick off the jobs
-		// to perform the pre-fetching (GH-1283).
-		defer func() {
-			rdr.err = group.Wait()
-			// don't close the last channel until after the group is finished,
-			// so that Next() can only return after reader.err may have been set
-			close(chs[lastChannelIndex])
-		}()
-
 		for i, b := range batches[1:] {
 			batch, batchIdx := b, i+1
 			chs[batchIdx] = make(chan arrow.Record, bufferSize)
@@ -611,6 +600,15 @@ func newRecordReader(ctx context.Context, alloc memory.Allocator, ld gosnowflake
 				return rr.Err()
 			})
 		}
+
+		// place this here so that we always clean up, but they can't be in a
+		// separate goroutine. Otherwise we'll have a race condition between
+		// the call to wait and the calls to group.Go to kick off the jobs
+		// to perform the pre-fetching (GH-1283).
+		rdr.err = group.Wait()
+		// don't close the last channel until after the group is finished,
+		// so that Next() can only return after reader.err may have been set
+		close(chs[lastChannelIndex])
 	}()
 
 	return rdr, nil

--- a/go/adbc/driver/snowflake/record_reader.go
+++ b/go/adbc/driver/snowflake/record_reader.go
@@ -567,6 +567,17 @@ func newRecordReader(ctx context.Context, alloc memory.Allocator, ld gosnowflake
 
 	lastChannelIndex := len(chs) - 1
 	go func() {
+		// place this here so that we always clean up, but they can't be in a
+		// separate goroutine. Otherwise we'll have a race condition between
+		// the call to wait and the calls to group.Go to kick off the jobs
+		// to perform the pre-fetching (GH-1283).
+		defer func() {
+			rdr.err = group.Wait()
+			// don't close the last channel until after the group is finished,
+			// so that Next() can only return after reader.err may have been set
+			close(chs[lastChannelIndex])
+		}()
+
 		for i, b := range batches[1:] {
 			batch, batchIdx := b, i+1
 			chs[batchIdx] = make(chan arrow.Record, bufferSize)
@@ -600,13 +611,6 @@ func newRecordReader(ctx context.Context, alloc memory.Allocator, ld gosnowflake
 				return rr.Err()
 			})
 		}
-	}()
-
-	go func() {
-		rdr.err = group.Wait()
-		// don't close the last channel until after the group is finished,
-		// so that Next() can only return after reader.err may have been set
-		close(chs[lastChannelIndex])
 	}()
 
 	return rdr, nil

